### PR TITLE
PrintingServices.Unix.cs thread safety, Issue #24268

### PR DESCRIPTION
**Pull request by Cojacfar**
Added locking to all writes to the Hashtable, as Hashtable supports multiple concurrent readers. Issue #24268

The getter is now this:
```c#
get
            {
                LoadPrinters();
                PrinterSettings.StringCollection list = new PrinterSettings.StringCollection(Array.Empty<string>());
                lock (installed_printers.SyncRoot)
                {

                    foreach (object key in installed_printers.Keys)
                    {
                        list.Add(key.ToString());
                    }
                    return list;
                }
            }

### CODE REVIEW

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 1]
```diff
@@ -75,11 +75,15 @@ internal static PrinterSettings.StringCollection InstalledPrinters
             {
                 LoadPrinters();
                 PrinterSettings.StringCollection list = new PrinterSettings.StringCollection(Array.Empty<string>());
-                foreach (object key in installed_printers.Keys)
+                lock (installed_printers.SyncRoot)
```
Comment (danmosemsft) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: Note that in .NET Core, SyncRoot for the major collections is just `this`
`        public virtual object SyncRoot => this;`
So you can just use `list` here and it's a little simpler.

**Comment by danmosemsft** [comment]
Some use of the hashtable is still missing a lock, eg., `return installed_printers.Contains(printer);` and also `installed_printers.Count`

For `if (installed_printers.Count == 0) LoadPrinters()` - although reading Count is thread safe in the sense it cannot corrupt the dictionary, it is not safe to read it, then take a lock and act as if it was zero, since an add could occur before you take the lock. The way LoadPrinters() is used, that may actually be safe, but it is fragile to use this kind of reasoning - it makes maintenance harder - unless it is for very pressing performance reasons, and then backed up with lots of comments.

There is another problem with the way the code is written - if there are no installed printers, it will try to load printers every time it is asked. 

I think you should
1) Put a lock around every access to the hashtable
2) Either move the count check into LoadPrinters, or better, add a boolean "populatedPrinters" and use that to prevent LoadPrinters being called more than once. That should eliminate any need to check Count.

**Comment by danmosemsft** [comment]
Thank you for the fix! I had overlooked this old bug and it will likely help stabilize our tests.

**Comment by Cojacfar** [comment]
Thanks for the feedback Dan. I'll put those updates in. I was thinking about the count problem after reading the other issue. This makes a lot of sense here to just lock it, since this is hardly a high-performance operation to check printers.

Also, just to double-check, the lock can be lock(installed_printers) with this change, nothing else. I'll get the new commit up here soon.

**Comment by danmosemsft** [comment]
> , the lock can be lock(installed_printers) with this change, nothing else

Yes that's right.

**Comment by Cojacfar** [comment]
I added populatedPrinters as a boolean variable, it's set after we add the printer to the Hashtable. I also have it using it's on lock object, but that might be overkill....

I don't think it makes sense to check for existing printers at the beginning of LoadPrinters because in theory you might just want to reload printers, as the first step is clearing out the hashtable.

**Comment by Cojacfar** [comment]
@danmosemsft I don't understand why this is failing two of the tests - it's failing corefx-ci because one of the tasks failed in Linux_x64 even though it's marked as completed right before. 

`/home/vsts_azpcontainer/.nuget/packages/microsoft.dotnet.helix.sdk/2.0.0-beta.19264.13/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : This task can only be used on completed jobs. There are 1 of 212 unfinished work items. [/__w/1/s/eng/sendtohelix.proj]`

**Comment by danmosemsft** [comment]
@dotnet/dnceng can you please help diagnose this: `Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(67,5): error : This task can only be used on completed jobs. There are 1 of 212 unfinished work items.`

cc @ViktorHofer

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 2]
```diff
@@ -589,7 +609,15 @@ private static PrinterResolution ParseResolution(string resolution)
         /// </summary>
         private static void LoadPrinters()
         {
-            installed_printers.Clear();
+            lock (installed_printers)
+            {
+                installed_printers.Clear();
+            }
+            lock (populatedLock)
```
Comment (danmosemsft) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: I don't think the boolean should have its own lock - also you have used it below (line 672) without a lock. I think the boolean should always be read and written within the `installed_printers` lock.

**Comment by ViktorHofer** [comment]
```
Executed on c000350 using docker image mcr.microsoft.com/dotnet-buildtools/prereqs:fedora-28-helix-09ca40b-20190508143249
running $HELIX_CORRELATION_PAYLOAD/scripts/aea2e2a4d1274413b501ee7ca7e3c12c/execute.sh in /home/helixbot/work/dbe7052a-a1f7-4445-b0a3-d61d3bb901f7/Work/bbadd3e3-f12f-4ab9-8e6d-f7e88e90e427/Exec max 900 seconds
```

That is a known issue and @MattGal is working on a fix.

**Comment by MattGal** [comment]
@danmosemsft the thing you asked about is https://github.com/dotnet/arcade/issues/2757 and the issue that caused the fireball as Viktor points out is tracked by https://github.com/dotnet/core-eng/issues/6327, specifically we need to wait until the docker daemon is responsive to start doing work on Linux.

**Comment by danmosemsft** [comment]
Thanks both!

**Comment by danmosemsft** [comment]
@safern could you help this old PR get in?

**Comment by karelz** [comment]
@safer @maryamariyan @Cojacfar what needs to happen to merge this PR?

**Comment by safern** [comment]
@Cojacfar it seems like there are some build failures would you mind fixing them to get this merged?

**Comment by maryamariyan** [comment]
Failures can be seen from here: https://dev.azure.com/dnceng/public/_build/results?buildId=291623

e.g.
```
System/Drawing/Printing/PrintingServices.Unix.cs(57,26): error SA1206: The 'static' modifier should appear before 'volatile'
Build failed (exit code '1').
Bash exited with code '1'.
```

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 3]
```diff
@@ -637,11 +658,15 @@ private static void LoadPrinters()
                             status = "Ready";
                             break;
                     }
-
-                    installed_printers.Add(name, new SysPrn.Printer(string.Empty, type, status, comment));
-
+                    lock (installed_printers)
+                    {
+                        installed_printers.Add(name, new SysPrn.Printer(string.Empty, type, status, comment));
+                        populatedPrinters = true;
```
[the remark this answers is not in the document]
Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: I want to say it's not a problem, but I see now where my thinking is wrong. I was envisioning polling the hashtable to pull out the printers, but in reality we pull from CUPS with load_printer every single time we retrieve available printers. Seems wasteful to me, but this isn't something I want to mess around with (how we interact with CUPS, I probably just don't understand it completely). 

I'll move the populated_printers flag setting outside of this for loop.

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 4]
```diff
@@ -589,7 +605,12 @@ private static PrinterResolution ParseResolution(string resolution)
         /// </summary>
         private static void LoadPrinters()
         {
-            installed_printers.Clear();
+            lock (installed_printers)
+            {
+                installed_printers.Clear();
+                populatedPrinters = false;
```
[the remark this answers is not in the document]
Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: @stephentoub Stephen, what do you think about the approach of making a task out of loadPrinter and storing this task with the static class. In this case, I will check if the task is currently running on each call and wait for it to end. Otherwise, I can start the task.

[the remark this answers is not in the document]
Comment (danmosemsft) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: @cojacfar I'm just arriving at this issue but this sounds like exactly what `Lazy<T>`  is designed for: https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization?view=netframework-4.8

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: > Stephen, what do you think about the approach

Seems like LazyInitializer.EnsureInitialized would suit this well.  Does it not?  Clearing would just be matter of setting the field to null, and EnsureInitialized would take a ref to the field and the delegate that constructs the list of printers and returns it.

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 5]
```diff
@@ -71,20 +72,35 @@ static PrintingServices()
...
+
+                    // Setting installed_printers back to null to cause EnsureInitialized to be called again after these threads finish
+                    // Also need this to dispose of the concurrently created hashtables
+                    installed_printers = null;
+
                     return list;
-                }
+                });
+
+
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: Extra line.

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 6]
```diff
@@ -71,20 +72,35 @@ static PrintingServices()
...
+        /// Returns a list of strings representing printers installed in CUPS
+        /// </summary>
         internal static PrinterSettings.StringCollection InstalledPrinters
         {
             get
             {
-                LoadPrinters();
-                PrinterSettings.StringCollection list = new PrinterSettings.StringCollection(Array.Empty<string>());
-                lock (installed_printers)
+                LazyInitializer.EnsureInitialized(ref installed_printers, () =>
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: will this work? `PrintinServices` has a static constructor which initializes `installed_printers` to a new Hashtable. So then I don't think the `LazyInitializer` will execute because it is already initialized.

Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: That's right, I must have not committed removing it.

Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: I guess changing the constructor creates a merge conflict, but none of the inputs changed. I believe this should work, but first time using LazyInitializer so let me know if it doesn't make sense.

Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: So I turned it into this: `LazyInitializer.EnsureInitialized(ref installed_printers, () =>
                {
                    installed_printers = new Hashtable();
                    // Loads values into installed_printers
                    LoadPrinters();
                    // Creates list from installed_printers that we return
                    PrinterSettings.StringCollection list = new PrinterSettings.StringCollection(Array.Empty<string>());
                    foreach (object key in installed_printers.Keys)
                    {
                        list.Add(key.ToString());
                    }
                    // Setting installed_printers back to null to cause EnsureInitialized to be called again after these threads finish
                    // Also need this to dispose of the concurrently created hashtables
                    installed_printers = null;
                    return list;
                });`

If I understand correctly, this should work. I had to drop the installed_printers creation in the constructor, which seems to have triggered a merge conflict. 

Each thread is going to make a new Hashtable, which is a downside. I could instead create a new property to just cause re-initialization. Thinking about it, that might be the best way.

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: So I turned it into this:

LazyInitializer.EnsureInitialized(ref installed_printers, () => { installed_printers = new Hashtable(); // Loads values into installed_printers LoadPrinters(); // Creates list from installed_printers that we return PrinterSettings.StringCollection list = new PrinterSettings.StringCollection(Array.Empty<string>()); foreach (object key in installed_printers.Keys) { list.Add(key.ToString()); } // Setting installed_printers back to null to cause EnsureInitialized to be called again after these threads finish // Also need this to dispose of the concurrently created hashtables installed_printers = null; return list; });

If I understand correctly, this should work. I had to drop the installed_printers creation in the constructor, which seems to have triggered a merge conflict.

Each thread is going to make a new Hashtable, which is a downside. I could instead create a new property to just cause re-initialization. Thinking about it, that might be the best way.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: I'm sorry if I mislead you somehow, but I don't understand what this is doing.  My suggestion was that the loading of the printers and the publishing of the relevant loaded data be done with LazyInitializer.EnsureInitialized, e.g.
```C#
private static Hashtable s_installedPrinters;

// When you need to read from the installed printers
Hashtable printers = LazyInitializer.EnsureInitialized(ref s_installedPrinters, () => 
{
    var printers = new Hashtable();
   ... // load all printers and store into printers
    return printers;
});
...
// When you need to clear the installed printers
Volatile.Write(ref s_installedPrinters, null);
...

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 7]
```diff
@@ -55,30 +56,43 @@ internal static class PrintingServices
         private static Hashtable installed_printers;
         private static string default_printer = string.Empty;
 
+        private static volatile bool populatedPrinters = false;
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: Why is this needed?

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 8]
```diff
@@ -55,30 +56,43 @@ internal static class PrintingServices
...
         #endregion
 
         #region Properties
 
+        private static Hashtable ReadPrinters()
+        {
+            var printers = new Hashtable();
+            printers = LoadPrinters(printers);
+            return printers;
+        }
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: Is LoadPrinters called from anywhere other than this method now?  Can we instead just make LoadPrinters create the Hashtable itself and return it, rather than taking one as an argument?  Then this ReadPrinters method isn't needed at all, and the uses of ReadPrinters can be replaced by LoadPrinters.

File: src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs [hunk 9]
```diff
@@ -587,9 +608,11 @@ private static PrinterResolution ParseResolution(string resolution)
 
         /// <summary>
         /// </summary>
-        private static void LoadPrinters()
+        private static Hashtable LoadPrinters(Hashtable printers)
         {
-            installed_printers.Clear();
+            Volatile.Write(ref installed_printers, null);
```
Comment (stephentoub) [inline src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: What's the purpose of this here?  This will effectively become a nop once the printers have been initialized, because LoadPrinters won't be called again.

In what situation / when do we need to reset and re-load?  This write should be moved to that location.

Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: In the original behavior we were calling LoadPrinters each time we requested the InstalledPrinters. I think this behavior is desired, otherwise how do we force the list of printers to update when CUPS does?

Comment (stephentoub) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: If the goal is to always have an up-to-date view, then why is there any caching at all?

Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: I was looking to resolve the multi-threading issue more than caching. I'm unsure of how to use the caching, as in theory at any time the system could add a new printer to CUPS. The caching between threads makes the most sense, and EnsureInitialized seems to do that. If I pass a lock object, I could have only one thread actually working instead of using the race condition. 

We also request the default printer, and I don't see a reason to re-run LoadPrinter there. But that was already taken care of with the PopulatedPrinters flag.

Comment (safern) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: I think what @stephentoub meant here is why do we want to cache the value and not every time populate/update the printers as it seems like we always want/need to have an up-to-date view. @stephentoub correct me if I'm wrong.

Comment (Cojacfar) [reply src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs]: Yeah, that makes sense to me. I was looking to use a single racing thread more than implement caching.

It would appear if I pass EnsureInitialized a sync object it will lock everything until the first thread finishes, then return a value. Otherwise, I can not pass an object and it will take the first thread that finishes.

I was thinking it's what you meant for it to be used for, more than the caching side from your comments earlier. 

> an optimistic concurrency scheme and allow multiple threads to race to build up results of which only one is published, I don't feel strongly about. The latter might be cleaner and less risky, though. You might be able to use LazyInitializer.EnsureInitialized to do it fairly cleanly.

**Comment by safern** [comment]
Thanks @Cojacfar for your work and patience here. Are you planning on updating the PR to address latest feedback? Let's try to get this merged before 11/13 so we don't have to move it to the new repo: dotnet/announcements#133

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

**Comment by maryamariyan** [comment]
hey @Cojacfar, a friendly reminder that this repo will be closed in two days (by Nov 13, 5pm PST) and all PRs in this repo will need to either be closed or merged by that date. 

As mentioned earlier you can still apply these changes and make a new PR to the new repo past this date by applying a patch of this diff. That's in case the PR doesn't get merged by then.

**Comment by stephentoub** [comment]
I pushed a commit to address my concerns.

@safern, @Cojacfar, can you please take a look at my changes?

Thanks.

**Comment by danmosemsft** [comment]
Thanks @Cojacfar for the contribution.
